rust/connlib/tunnel/src/tests/reference.rs

```diff
@@ -172,37 +178,127 @@ impl ReferenceStateMachine for ReferenceState {
+                )
+                .boxed(),
+            ));
         }
 
-        if !state.resolved_ips_for_non_resources().is_empty() {
-            strategies.push((1, icmp_to_resolved_non_resource().boxed()));
+        let dns_v4_domains = state.resolved_v4_domains();
+        if !dns_v4_domains.is_empty() {
+            strategies.push((
+                1,
+                icmp_to_dns_resource(
+                    packet_source_v4(state.client.tunnel_ip4).prop_map(IpAddr::V4),
+                    sample::select(dns_v4_domains),
+                )
+                .boxed(),
+            ));
+        }
+
+        let dns_v6_domains = state.resolved_v6_domains();
+        if !dns_v6_domains.is_empty() {
+            strategies.push((
+                1,
+                icmp_to_dns_resource(
+                    packet_source_v6(state.client.tunnel_ip6).prop_map(IpAddr::V6),
+                    sample::select(dns_v6_domains),
+                )
+                .boxed(),
+            ));
+        }
+
+        let domains = state.all_domains();
+        let v4_dns_servers = state.v4_dns_servers();
+        let v6_dns_servers = state.v6_dns_servers();
+
+        if !domains.is_empty()
+            && !state.v4_dns_servers().is_empty()
+            && state.client.ip4_socket.is_some()
+        {
+            strategies.push((
+                1,
+                dns_query(
+                    sample::select(domains.clone()),
+                    sample::select(v4_dns_servers).prop_map(SocketAddr::V4),
+                )
+                .boxed(),
+            ));
+        }
+
+        if !domains.is_empty()
+            && !state.v6_dns_servers().is_empty()
+            && state.client.ip6_socket.is_some()
+        {
+            strategies.push((
+                1,
+                dns_query(
+                    sample::select(domains),
+                    sample::select(v6_dns_servers).prop_map(SocketAddr::V6),
+                )
+                .boxed(),
+            ));
+        }
+
+        let resolved_non_resource_ip4s = state.resolved_ip4_for_non_resources();
+        if !resolved_non_resource_ip4s.is_empty() {
+            strategies.push((
+                1,
+                ping_random_ip(
+                    packet_source_v4(state.client.tunnel_ip4).prop_map(IpAddr::V4),
+                    sample::select(resolved_non_resource_ip4s).prop_map(IpAddr::V4),
+                )
+                .boxed(),
+            ));
+        }
+
+        let resolved_non_resource_ip6s = state.resolved_ip6_for_non_resources();
+        if !resolved_non_resource_ip6s.is_empty() {
+            strategies.push((
+                1,
+                ping_random_ip(
+                    packet_source_v6(state.client.tunnel_ip6).prop_map(IpAddr::V6),
+                    sample::select(resolved_non_resource_ip6s).prop_map(IpAddr::V6),
+                )
+                .boxed(),
+            ));
```

Comment:
I can try to refactor something here. It is a bit of a trade-off. I don't want to extract functions for creating the actual strategies because they are all compositions of more generic strategies.

I can try hiding the `if` and see if that makes things better?

Reply:
Let me know what you think of https://github.com/firezone/firezone/pull/5350.